### PR TITLE
ENH: Add support for platforms with missing fenv flags

### PR DESCRIPTION
**Pull request by hoodmane**
For instance wasm has no fenv support. musl-libc defines all the
fenv functions unconditionally (fesetexcept, feclearexcept, etc)
but on unsupported platforms they are no-ops:
https://git.musl-libc.org/cgit/musl/tree/src/fenv/fenv.c

However, the platform is expected to only define flags like
FE_OVERFLOW if they are supported. I haven't found an explanation
of the design decision, but it seems to be aimed at fine-grained
feature detection. musl-libc-test has code that uses fenv and wants
it to silently decay to a no-op on platforms missing support. I
copied their implementation of this behavior:
http://nsz.repo.hu/git/?p=libc-test;a=blob;f=src/common/mtest.h;h=706c1ba23ea8989b17a2f72ed1a919e187c06b6a;hb=HEAD#l30

### CODE REVIEW

**Comment by seberg** [comment]
I suppose we can just do this.  But I guess the NumPy test suite will complain on a platform without any FPE support?

**Comment by hoodmane** [comment]
You can try it out:
1. Go to https://pyodide-cdn2.iodide.io/dev/full/console.html
2. open browser console
3. paste and run the following:
```py
await pyodide.loadPackage(["numpy-tests","pytest", "micropip"]);
await pyodide.pyimport("micropip").install("hypothesis");
pyodide.runPython(`
import pytest
pytest.main(["--pyargs", "numpy", "-vv"]) 
`);
```

**Comment by hoodmane** [comment]
I don't see any test failures so far but I am only part way through.

**Comment by hoodmane** [comment]
Okay `core/tests/test_numeric.py::TestFloatExceptions` all fail.

**Comment by seberg** [comment]
I think we have other platforms that we at least somewhat support, that have e.g. software floating points only and thus have similar problems, so some tests might be written to be OK.
In general we should just merge this, though (modulo possible style nitpicks).

Two possible followups (but low priority IMO):
* Make the test suite pass on such system
* If the test suite passes, maybe we should consider giving users some indication of floating point support?  If just printing it out in the test-suite `np.test()`.

**Comment by hoodmane** [comment]
The results of running the full test suite are:
```
= 194 failed, 16807 passed, 1355 skipped, 23 xfailed, 2 xpassed, 230 warnings, 10 errors in 529.46s (0:08:49) =
```
Could be a lot worse.

File: numpy/core/src/npymath/ieee754.c.src [hunk 1]
```diff
@@ -789,30 +789,72 @@ int npy_clear_floatstatus_barrier(char *param)
...
+#ifdef FE_UNDERFLOW
+#define UNDERFLOW FE_UNDERFLOW
+#else
+#define UNDERFLOW 0
+#endif
+#ifdef FE_OVERFLOW
+#define OVERFLOW FE_OVERFLOW
+#else
+#define OVERFLOW 0
+#endif
```
Comment (seberg) [inline numpy/core/src/npymath/ieee754.c.src:831]: This is a local C-file, any reason not to shorten this to?
```C
#ifndef FE_INEXACT
    #define FE_INEXACT 0
#endif
```
I might also prefer shorting the comment a bit:
```C
/*
 * According to the C99 standard FE_INEXACT, etc. may
 * not be provided when not supported.
 * In such cases NumPy will not report these correctly,
 * but we should still allow compiling making all FP error
 * handling effectively no-ops.
 */
```
(Maybe continue with a reference to musl?)

Anyway, super nitpicky, so am happy to just fix it up myself or keep it if you think it is no improvement.

Comment (hoodmane) [reply numpy/core/src/npymath/ieee754.c.src:831]: IMO the two musl links should stay since they are probably helpful. Other than that your reworded comment seems good to me.

**Comment by seberg** [comment]
Just ran the NumPy 1.8.1 test-suite on this.  It has 16 (related) failures, out of which 15 are all in the same `test_distances` file and seem to fail for the same `integer_arr[something] = integer_arr[something] * np.nan` assignment.  So I suspect they should just deal with integer arrays better (especially since that assignment may be undefined behavior for some integer dtypes!).

**Comment by seberg** [comment]
Sorry, that went to the wrong PR (as for this one, I will update it next and merge it tomorrow if no-one has a comment).

**Comment by seberg** [comment]
I am a bit confused here.  We currently have a `.c.src` and a `.cpp` version of the file.  I guess we want to keep both around, because having to go back to the `.c.src` version is just a temporary embarrassment ;)?
The move to `.cpp` also rebased away gh-21421, so I want to fix that up first (after I figured out whether I should fix both c and cpp, or not).

**Comment by seberg** [comment]
Thanks @hoodmane, if anyone doesn't like my fix-ups please complain :).

**Comment by hoodmane** [comment]
Thanks @seberg!
